cpp/log/cert.cc

```diff
@@ -43,6 +43,100 @@ static int X509_get_signature_nid(const X509* x) {
 #endif
 
 
+namespace {
+#if defined(OPENSSL_IS_BORINGSSL)
+// BoringSSL doesn't have DSA hooked up so to accept these certs we have
+// to do the work ourselves. Can be called with a non DSA issuer key but will
+// always return false in that case.
+//
+// cert is the certificate that is to be checked
+// issuer_key is the public key of the certificate issuer (should be DSA)
+// Returns true if the DSA signature was correctly verified or false if
+// it did not or there was any error along the way.
+StatusOr<bool> check_dsa_signature(const X509* cert,
+                                   EVP_PKEY* issuer_key) {
+  // Before we start rummaging about through pointers ensure everything we need
+  // is present
+  if (!cert || !issuer_key || !X509_get_cert_info(cert)) {
+    return ::util::Status(Code::FAILED_PRECONDITION,
+                          "cert is null or missing cert_info");
+  }
+
+  X509_CINF* cert_info = X509_get_cert_info(cert);
+  const X509_ALGOR* sig = X509_CINF_get_signature(cert_info);
+
+  if (!sig) {
+    return ::util::Status(Code::FAILED_PRECONDITION,
+                          "cert is missing a signature");
+  }
+
+  if (EVP_PKEY_type(issuer_key->type) != EVP_PKEY_DSA) {
+    return ::util::Status(Code::FAILED_PRECONDITION,
+                          "issuer does not have a DSA public key");
+  }
+
+  const int alg_nid = OBJ_obj2nid(sig->algorithm);
+
+  int digest_nid;
+  // We need the nid of the digest so we can create an EVP_MD for it later.
+  // Should succeed as we already checked we have a DSA key
+  if (!OBJ_find_sigid_algs(alg_nid, &digest_nid, nullptr)) {
+    return ::util::Status(Code::INTERNAL, "lookup sigid for algorithm failed");
+  }
+
+  // Get the DER encoded certificate info from the cert
+  unsigned char* der_buf(nullptr);
+  const int der_length = i2d_X509_CINF(cert_info, &der_buf);
+  if (der_length < 0) {
+    // What does this return value mean? Let's assume it means the cert
+    // is bad until proven otherwise.
```

Comment:
I've replaced all occurrences of this comment with a note that the decode / serialize failed and we're rejecting the input rather than trying to work out exactly what happened.